In [28]:
#Neccessary Librabries 
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import confusion_matrix
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge, Lasso, ElasticNet

In [29]:
df=pd.read_csv("C:/Users/pvsai/Downloads/archive (8)/realest.csv")

In [30]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Price      156 non-null    float64
 1   Bedroom    156 non-null    float64
 2   Space      146 non-null    float64
 3   Room       156 non-null    float64
 4   Lot        146 non-null    float64
 5   Tax        147 non-null    float64
 6   Bathroom   156 non-null    float64
 7   Garage     156 non-null    float64
 8   Condition  156 non-null    float64
dtypes: float64(9)
memory usage: 11.2 KB


Price     Bedroom        Space        Room         Lot  \
count  156.000000  156.000000   146.000000  156.000000  146.000000   
mean    56.474359    3.166667  1097.246575    6.500000   32.808219   
std     12.875307    1.348037   462.540698    1.675247    8.457859   
min     32.000000    1.000000   539.000000    4.000000   24.000000   
25%     46.000000    2.000000   805.250000    5.000000   25.000000   
50%     55.000000    3.000000   965.500000    6.000000   30.000000   
75%     65.000000    4.000000  1220.500000    7.000000   37.000000   
max     90.000000    8.000000  2295.000000   12.000000   50.000000   

               Tax    Bathroom      Garage   Condition  
count   147.000000  156.000000  156.000000  156.000000  
mean    911.707483    1.480769    0.846154    0.230769  
std     443.263430    0.529408    0.808454    0.422682  
min     418.000000    1.000000    0.000000    0.000000  
25%     652.500000    1.000000    0.000000    0.000000  
50%     821.000000    1.500000    1.000000    0.000000  
75%    1012.500000    2.000000    1.500000    0.000000  
max    2752.000000    3.000000    2.000000    1.000000

In [31]:
df.fillna(df.median(), inplace=True)
missing_values = df.isnull().sum()
missing_values

Price        0
Bedroom      0
Space        0
Room         0
Lot          0
Tax          0
Bathroom     0
Garage       0
Condition    0
dtype: int64

In [33]:
# 1. Regression Models to predict the prices of the house
# Split data for regression into training(80%) and testing(20%).
Reg = df.drop(columns='Price')
Reg2 = df['Price']
Reg_train, Reg_test, Reg2_train, Reg2_test = train_test_split(Reg, Reg2, test_size=0.2, random_state=42)

In [34]:
# Linear Regression
# Initialize and train the Linear Regression model
lr = LinearRegression()
lr.fit(Reg_train, Reg2_train)

# Predict on the testing set
Reg2_pred_lr = lr.predict(Reg_test)

# Evaluate the Linear Regression model
print("Linear Regression Mean Squared Error:", mean_squared_error(Reg2_test, Reg2_pred_lr))
print("Linear Regression R^2 Score:", r2_score(Reg2_test, Reg2_pred_lr))


Linear Regression Mean Squared Error: 67.61406527336344
Linear Regression R^2 Score: 0.5996113735518254


In [36]:
# Random Forest Regressosion
# Initialize and train the Random Forest Regressosion
rfr = RandomForestRegressor(n_estimators=1000, random_state=42)
rfr.fit(Reg_train, Reg2_train)

# Predict on the testing set
Reg2_pred_rfr = rfr.predict(Reg_test)

# Evaluate the Random Forest Regressor
print("Random Forest Regressor Mean Squared Error:", mean_squared_error(Reg2_test, Reg2_pred_rfr))
print("Random Forest Regressor R^2 Score:", r2_score(Reg2_test, Reg2_pred_rfr))


Random Forest Regressor Mean Squared Error: 24.674909656249998
Random Forest Regressor R^2 Score: 0.853883165506234


In [37]:
# Gradient Boosting Regressosion
# Initialize and train the Random Forest Regressosion
gbr = GradientBoostingRegressor(n_estimators=1000, learning_rate=0.1, max_depth=3, random_state=42)
gbr.fit(Reg_train, Reg2_train)

# Predict on the testing set
Reg2_pred_gbr = gbr.predict(Reg_test)

# Evaluate the Gradient Boosting Regressor
print("Gradient Boosting Regressor Mean Squared Error:", mean_squared_error(Reg2_test, Reg2_pred_gbr))
print("Gradient Boosting Regressor R^2 Score:", r2_score(Reg2_test, Reg2_pred_gbr))

Gradient Boosting Regressor Mean Squared Error: 22.741056032594994
Gradient Boosting Regressor R^2 Score: 0.8653348212082923


In [38]:
# L2 Regularization: Ridge Regression
ridge = Ridge(alpha=1.0)
ridge.fit(Reg_train, Reg2_train)
ridge_pred = ridge.predict(Reg_test)
print("Ridge Mean Squared Error:", mean_squared_error(Reg2_test, ridge_pred))
print("Ridge R^2 Score:", r2_score(Reg2_test, ridge_pred))

# L1 Regularization: Lasso Regression
lasso = Lasso(alpha=0.1)
lasso.fit(Reg_train, Reg2_train)
lasso_pred = lasso.predict(Reg_test)
print("Lasso Mean Squared Error:", mean_squared_error(Reg2_test, lasso_pred))
print("Lasso R^2 Score:", r2_score(Reg2_test, lasso_pred))

# L1 + L2 Regularization: Elastic Net
elastic= ElasticNet(alpha=0.1, l1_ratio=0.5)
elastic.fit(Reg_train, Reg2_train)
elastic_pred = elastic.predict(Reg_test)
print("Elastic Net Mean Squared Error:", mean_squared_error(Reg2_test, elastic_pred))
print("Elastic Net R^2 Score:", r2_score(Reg2_test, elastic_pred))

Ridge Mean Squared Error: 67.56420331371083
Ridge R^2 Score: 0.5999066399502677
Lasso Mean Squared Error: 68.00815731972337
Lasso R^2 Score: 0.5972776879126279
Elastic Net Mean Squared Error: 67.66962221245691
Elastic Net R^2 Score: 0.5992823833270345


# Conclusion:
The R^2 scores are:
Linear Regression: 0.59
Random Forest Regressor: 0.85
Gradient Boosting Regressor: 0.86
Ridge: 0.59
Lasso: 0.59
Elastic Net: 0.59

In Regression model, Gradient Boosting Regressor is the best model with R^2 score of 0.86.